lingua_franca/internal.py

```diff
@@ -0,0 +1,503 @@
+        return lang
+    elif lang in _DEFAULT_FULL_LANG_CODES:
+        return _DEFAULT_FULL_LANG_CODES[lang]
+    else:
+        return "en-us"
+
+
+def localized_function(run_own_code_on=[type(None)]):
+    """
+    Decorator which finds localized functions, and calls them, from signatures
+    defined in the top-level modules. See lingua_franca.format or .parse for
+    examples of the decorator in action.
+
+    Note that, by default, wrapped functions will never actually be executed.
+    Rather, when they're called, their arguments will be passed directly to
+    their localized equivalent, specified by the 'lang' parameter.
+
+    The wrapper can be instructed to execute the wrapped function itself when
+    a specified error is raised (see the argument 'run_own_code_on')
+
+    For instance, this decorator wraps parse.extract_number(), which has no
+    logic of its own. A call to
+
+        extract_number('uno', lang='es')
+
+    will locate and call
+
+        lingua_franca.lang.parse_es.extract_number_es('uno')
+
+    By contrast, here's the decorator above format.nice_number, with the param:
+
+        @localized_function(run_own_code_on=[UnsupportedLanguageError])
+        def nice_number(number, lang=None, speech=True, denominators=None):
+
+    Here, nice_number() itself will be executed in the event that the localizer
+    raises an UnsupportedLanguageError.
+
+    Arguments:
+        run_own_code_on(list(type), optional)
+            A list of Error types (ValueError, NotImplementedError, etc)
+            which, if they are raised, will trigger the wrapped function's
+            own code.
+
+            If this argument is omitted, the function itself will never
+            be run. Calls to the wrapped function will be passed to the
+            appropriate, localized function.
+
+
+    """
+    # Make sure everything in run_own_code_on is an Error or None
+    def is_error_type(_type):
+        _instance = _type()
+        rval = isinstance(_instance, BaseException) if _instance else True
+        del _instance
+        return rval
+    if not isinstance(run_own_code_on, list):
+        run_own_code_on = list(run_own_code_on)
+    if run_own_code_on != [None]:
+        if not all((is_error_type(e) for e in run_own_code_on)):
+            raise ValueError("@localized_function(run_own_code_on=<>) expected an "
+                             "Error type, or a list of Error types. Instead, it "
+                             "received this value:\n" + str(run_own_code_on))
+
+    # Begin wrapper
+    def localized_function_decorator(func):
+        @wraps(func)
+        def call_localized_function(*args, **kwargs):
+            try:
+                func_signature = signature(func)
+                func_params = list(func_signature.parameters)
+                lang_param_index = func_params.index('lang')
+
+                lang_code = get_default_lang()
+                full_lang_code = __active_lang_code
```

Comment:
Here, `lang_code` is reassigned in the event that a `lang=` parameter is passed. The quirk, I suppose, is just the order in which I've done the assignments:

1. Assign `lang_code` to the current default. If there is no default, this is `None`
1. If there's a keyword argument `lang`, overwrite it with that value
1. If not, look for a positional argument, and overwrite it with that
1. If `lang_code` is still `None`, raise `ModuleNotFoundError: No language module loaded.`

Reply:
sorry to clarify, I was talking about the assignment of `full_lang_code`

Reply:
Gotcha! That's also overridden if we find another lang_code other than the default. However, and this is a great catch, the combination of the two can override a custom full code (like it might kick 'en-au' back to 'en-us') so that's something to fix!